# HG changeset patch
# Parent 2741c6558064451a7ff52d66897fa3f3fe17ef81
# User Ben Kelly <ben@wanderview.com>
Mody Fetch to wrap non-buffered streams in a new nsBufferedInputStream.


diff --git a/dom/fetch/Fetch.cpp b/dom/fetch/Fetch.cpp
--- a/dom/fetch/Fetch.cpp
+++ b/dom/fetch/Fetch.cpp
@@ -697,17 +697,27 @@ FillUTF8Decoded(nsIInputStream* in,
 
 template <class Derived>
 void
 FetchBody<Derived>::FinishConsumeBody()
 {
   nsCOMPtr<nsIInputStream> stream;
   DerivedClass()->GetBody(getter_AddRefs(stream));
   MOZ_ASSERT(stream);
-  MOZ_ASSERT(NS_InputStreamIsBuffered(stream));
+
+  if (!NS_InputStreamIsBuffered(stream)) {
+    nsCOMPtr<nsIInputStream> buffered;
+    nsresult rv = NS_NewBufferedInputStream(getter_AddRefs(buffered),
+                                            stream, 4096);
+    if (NS_WARN_IF(NS_FAILED(rv))) {
+      return mDelayedConsumePromise->MaybeReject(rv);
+    }
+
+    stream = buffered.forget();
+  }
 
   uint64_t len;
   nsresult rv = stream->Available(&len);
   if (NS_WARN_IF(NS_FAILED(rv))) {
     return mDelayedConsumePromise->MaybeReject(rv);
   }
 
   AutoJSAPI api;
